pkg/bgp/controller/subscriber.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"github.com/cilium/cilium/pkg/k8s"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+
+	"go.universe.tf/metallb/pkg/k8s/types"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// NewSubscriber creates a subscriber of service objects for the BGP.
+func NewSubscriber(c *Controller, swgSvcs, swgEps *lock.StoppableWaitGroup, indexer cache.Store) *Subscriber {
+	s := &Subscriber{
+		c:       c,
+		swgSvcs: swgSvcs,
+		swgEps:  swgEps,
+
+		queue:   workqueue.New(),
+		indexer: indexer,
+	}
+	go s.run()
+	return s
+}
+
+// Subscriber subscribes and handles K8s services events from the watcher and
+// pushes them into a queue. From the queue, they are processed by the
+// reconciliation logic of the BGP controller via the MetalLB logic for LB IP
+// allocation. This handler implements
+// github.com/cilium/cilium/pkg/k8s/watchers/subscriber.ServiceHandler and
+// therefore is registered as a subscriber to the subscriber package to be
+// called from the K8s watcher.
+//
+// The allocation occurs only for services of type LoadBalancer in the
+// service.Status.LoadBalancerStatus.Ingress field.
+type Subscriber struct {
+	c               *Controller
+	swgSvcs, swgEps *lock.StoppableWaitGroup
+
+	// queue holds all services that need to be reconciled.
+	queue workqueue.Interface
+	// indexer is the store containing all the slim_corev1.Service objects seen
+	// by the controller. This is used in order to handle delete events. See
+	// comment inside (*Subscriber).run().
+	indexer cache.Store
+}
+
+// OnAdd handles an add event for services. It implements
+// github.com/cilium/cilium/pkg/k8s/watchers/subscriber.ServiceHandler.
+func (s *Subscriber) OnAdd(obj *slim_corev1.Service) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		s.queue.Add(svcEvent(key))
+	}
```

Comment:
I've added simple warnings similar to what we do inside `pkg/k8s/factory_functions.go` when a K8s object is not how we expect it. 

In terms of adding more logging throughout, MetalLB code already logs events internally. However I've decided to integrate those logs into Cilium only when debug is enabled, because they can be very chatty. I think as a followup, we should consider more deeply how one might introspect the state of BGP in Cilium without having to turn on debugging & look at chatty logs. I've added that as a followup item because MetalLB itself doesn't provide much itself (https://metallb.universe.tf/configuration/troubleshooting/), so we'd have to implement this ability ourselves.